R/ordering.R

```diff
@@ -82,3 +82,90 @@ setMethod("ordering<-", "DatasetCatalog", function (x, value) {
+#' }
+moveToAfter <- function(x, value, after){
+    if (!inherits(after, "OrderGroup")) {
+        after <- urls(after)
+    }
+    if (!inherits(value, "OrderGroup")) {
+        value <- urls(value)
+    }
+    whi <- ifelse(inherits(after, "OrderGroup"), which(names(x) %in% name(after)), which(entities(x) %in% after))
+    if ((inherits(value, "OrderGroup") && !name(value) %in% names(x)) || (!inherits(value, "OrderGroup") && !value %in% entities(x))) entities(x) <- c(entities(x), value)
+    whi2 <- ifelse(inherits(value, "OrderGroup"), which(names(x) %in% name(value)), which(entities(x) %in% value))
+    entities(x) <- entities(x)[c(setdiff(1:whi, whi2), whi2, setdiff((whi+1):length(entities(x)), whi2))]
+    return(x)
+}
+
+#' Copy the order from one dataset to another.
+#'
+#' @param source the dataset you want to copy the order from
+#' @param target the dataset you want to copy the order to
+#' @return returns an object of class [`VariableOrder`] (which can be assigned
+#' to a dataset with [`ordering`])
+#' @examples
+#' \dontrun{
+#' ordering(ds) <- copyOrder(ds1, ds)
+#' }
+#' @export
+copyOrder <- function(source, target){
+    if (!is.dataset(source) | !is.dataset(target)) {
+        halt("Both source and target must be Crunch datasets.")
+    }
+
+    ord <- entities(ordering(source))
+    new_ord <- lapply(ord, copyOrderGroup,  source = source, target = target)
+
+    # TODO: copy other VariableOrder features?
+
+    return(do.call(VariableOrder, new_ord))
+}
+
+#' Copy the order of a `VariableGroup` (or individual variable url) from `VariableOrder`
+#'
+#' @param gr the group or variable url to be copied
+#' @param source the dataset you want to copy the order from
+#' @param target the dataset you want to copy the order to
+#' @return returns either a [`VariableGroup`] (if a group is supplied) or a url (if just a variable url is supplied)
+#' @examples
+#' \dontrun{
+#' ordering(ds) <- copyOrder(ds1, ds)
+#' }
+#' @keywords internal
+copyOrderGroup <- function(gr, source, target) {
+    # if there is a single element in gr, and it is a character,
+    # just return the URL in the target.
+    if (length(gr) == 1 & is.character(gr)) {
+        al <- aliasFromURL(gr, source)
+        return(urls(target[[al[al %in% aliases(allVariables(target))]]]))
+    }
+
+    # make mask of which elements of gr are groups and which are variables
+    groups <- vapply(gr, class, character(1)) == "VariableGroup"
+    # make an empty list the same length as gr
+    ents <- vector("list", length(gr))
+
+    # deal with non-groups (if there are no non-groups, nothing will happen here)
+    als <- aliasFromURL(entities(gr[!groups]), source)
+    ents[!groups] <- urls(target[als[als %in% aliases(allVariables(target))]])
+
+    # deal with groups (if there are no groups, nothing will happen here)
+    ents[groups] <- lapply(entities(gr[groups]), copyOrderGroup,  source = source, target = target)
+
+    return(VariableGroup(name(gr), ents))
```

Comment:
Thanks, I was messing `removeMissingEntities` I've actually had to include it twice since both `VariableGroup` and `VariableOrder` fail if given `NULL` or `NA` entries. I've simplified  `copyOrderGroup` a bunch since it traverses until it gets to a single character — this isn't exactly the same pattern as `removeMissingEntities` but is close (and much simpler than what was there before!)